crates/accelerate/src/convert_2q_block_matrix.rs

```diff
@@ -0,0 +1,72 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use pyo3::Python;
+
+use num_complex::Complex64;
+use numpy::ndarray::linalg::kron;
+use numpy::ndarray::{Array, Array2, ArrayView2};
+use numpy::{IntoPyArray, PyArray2, PyReadonlyArray2};
+
+/// Return the matrix Operator resulting from a block of Instructions.
+#[pyfunction]
+#[pyo3(text_signature = "(op_list, /")]
+pub fn blocks_to_matrix(
+    py: Python,
+    op_list: Vec<(PyReadonlyArray2<Complex64>, Vec<usize>)>,
+) -> PyResult<Py<PyArray2<Complex64>>> {
+    let identity: Array2<Complex64> = Array::eye(2);
+    let input_matrix = op_list[0].0.as_array();
+    let mut matrix: Array2<Complex64> = match op_list[0].1.as_slice() {
+        [0] => kron(&identity, &input_matrix),
+        [1] => kron(&input_matrix, &identity),
+        [0, 1] => input_matrix.to_owned(),
+        [1, 0] => change_basis(input_matrix),
+        _ => unreachable!(),
+    };
+    for (op_matrix, q_list) in op_list.into_iter().skip(1) {
+        let op_matrix = op_matrix.as_array();
+        let q_list = q_list.as_slice();
+        let result = match q_list {
+            [0] => Some(kron(&identity, &op_matrix)),
+            [1] => Some(kron(&op_matrix, &identity)),
+            [1, 0] => Some(change_basis(op_matrix)),
+            _ => None,
+        };
+        matrix = match result {
+            Some(result) => result.dot(&matrix),
+            None => op_matrix.dot(&matrix),
+        };
+    }
+    Ok(matrix.into_pyarray(py).to_owned())
+}
+
+/// Switches the order of qubits in a two qubit operation.
+fn change_basis(matrix: ArrayView2<Complex64>) -> Array2<Complex64> {
+    let mut trans_matrix: Array2<Complex64> = matrix.reversed_axes().to_owned();
+    for index in 0..trans_matrix.ncols() {
+        trans_matrix.swap([1, index], [2, index]);
+    }
+    trans_matrix = trans_matrix.reversed_axes();
+    for index in 0..trans_matrix.ncols() {
+        trans_matrix.swap([1, index], [2, index]);
+    }
+    trans_matrix
+}
```

Comment:
This clones the matrix, but we'd be able to mutate the one we're given if we took in an `ArrayViewMut2`, since we never need it in the other form.

Reply:
Well, there's a problem with doing so, since we're receiving `PyReadOnlyArray` from Python, which doesn't implement the `DataMut` trait that's necessary to turn the views to mutable views by calling `view_mut()`.
One way of avoiding that is by using the `as_array_mut()` method to obtain an `ArrayViewMut2` right away. However, this is an unsafe operation, and I'm not sure we'd want to include unsafe code for this solution.

Reply:
Yeah, I was kind of meaning that we'd take in a `PyReadwriteArray2` at the top level instead, but really, I wouldn't worry about it - I think it'd need fairly significantly changes to how we're handling the iteration here to make it borrow-safe.